### PR TITLE
Add ~/.history file for Shell

### LINKED ISSUE #174: Shell: Command history file
It would be cool if Shell would save all the executed commands to `~/.history` or some similarly-named file.
The history should also be loaded on Shell start-up so we can immediately press "up" and see commands from an earlier session.

### PR DESCRIPTION
**Pull request by CallumAttryde**
Added a history file for the Shell, which fixes #174. Saves every executed command to the file, loads the last n capacity commands into the history buffer for each new shell (seeking from near the end of the file). File is saved to /home/[user]/.history

### CODE REVIEW

File: Shell/main.cpp [hunk 1]
```diff
@@ -561,6 +561,17 @@ int main(int argc, char** argv)
         endpwent();
     }
 
+    StringBuilder sb;
+    sb.append(g.home);
+    sb.append("/.history");
+    CFile history_file(sb.to_string());
+    rc = history_file.open(CIODevice::ReadWrite);
+    if (rc < 0)
```
Comment (awesomekling) [inline Shell/main.cpp]: `CIODevice::open()` returns a `bool`, so the appropriate error value to check for is `false`.

**Comment by CallumAttryde** [comment]
Isn't there a potential for commands to get lost with that design? For example, user enters 110 commands into the shell, the first 10 of which get popped off the history buffer and so never persisted. I suppose that wouldn't matter for this issue assuming the history buffer capacity never changed, but it might cause spooky behaviour with potential future features such as reverse-i-search

Ok, I implemented that to alleviate the problem of ever growing history files causing each new Shell to take longer to load up. I'll scan the whole file if that isn't an issue

**Comment by awesomekling** [comment]
That's okay, we would already have lost them due to the capacity setting. Ultimately we need to have *some* cap. 100 is probably way too small though :)

**Comment by CallumAttryde** [comment]
Changed the implementation to reflect review comments. There is an issue where atexit doesn't appear to be called when the Shell is closed via the Quit action on the Terminal. I couldn't find the bug in the Terminal/GApplication code, so not sure how to fix

File: Shell/main.cpp [hunk 2]
```diff
@@ -528,6 +528,36 @@ static int run_command(const String& cmd)
...
 }
 
+CFile get_history_file()
+{
+    StringBuilder sb;
+    sb.append(g.home);
+    sb.append("/.history");
+    CFile f(sb.to_string());
+    if (!f.open(CIODevice::ReadWrite))
+        perror("history file");
```
Comment (awesomekling) [inline Shell/main.cpp]: `perror()` will print whatever is in `errno`. There's no guarantee that it's relevant to the failed `open` call.
Instead, use `CFile::error_string()` to find the error message that relates to the failure.

**Comment by DavidMichaelEzzell** [comment]
Not happening today

**Comment by awesomekling** [comment]
This is pretty cool, thanks for implementing it @CallumAttryde :)

**Pull request merged**